corehq/apps/case_search/tests/test_filter_dsl.py

```diff
@@ -22,169 +23,376 @@
-                                {
-                                    "term": {
-                                        "case_properties.value.exact": -100.32,
+                                    "bool": {
+                                        "filter": (
+                                            {
+                                                "term": {
+                                                    "case_properties.key.exact": "number"
+                                                }
+                                            },
+                                            {
+                                                "term": {
+                                                    "case_properties.value.exact": -100.32
+                                                }
+                                            }
+                                        )
                                     }
                                 }
-                            )
+                            ],
+                            "must": {
+                                "match_all": {}
+                            }
+                        }
+                    }
+                }
+            }
+        else:
+            expected_filter = {
+                "nested": {
+                    "path": "case_properties",
+                    "query": {
+                        "filtered": {
+                            "query": {
+                                "match_all": {}
+                            },
+                            "filter": {
+                                "and": (
+                                    {
+                                        "term": {
+                                            "case_properties.key.exact": "number"
+                                        }
+                                    },
+                                    {
+                                        "term": {
+                                            "case_properties.value.exact": -100.32,
+                                        }
+                                    }
+                                )
+                            }
                         }
                     }
                 }
             }
-        }
         built_filter = build_filter_from_ast("domain", parsed)
-        self.assertEqual(expected_filter, built_filter)
+        self.checkQuery(expected_filter, built_filter, is_raw_query=True)
 
     def test_case_property_existence(self):
         parsed = parse_xpath("property != ''")
-        expected_filter = {
-            "and": (
-                {
-
-                    "nested": {
-                        "path": "case_properties",
-                        "query": {
-                            "filtered": {
-                                "query": {
-                                    "match_all": {
+        if self.is_es7():
+            expected_filter = {
```

Comment:
@sravfeyn the ES7 and ES2 versions of this query, while logically equivalent, are quite different. This is my reading of what they mean:

ES7: NOT (NOT property is set OR property  == "")
ES2: property is set AND NOT property == ""

I do see the ES7 version is produced by ``

https://github.com/dimagi/commcare-hq/blob/c5d65a8ff06e0b84819ea8378efa1cc5fb330500/corehq/apps/case_search/filter_dsl.py#L179-L185

and `case_property_missing`

https://github.com/dimagi/commcare-hq/blob/c5d65a8ff06e0b84819ea8378efa1cc5fb330500/corehq/apps/es/case_search.py#L245-L256

but I am not seeing how or where the ES2 version of this query is produced. Do you know where that's coming from or is there a bug in this test?

Reply:
oooh, I think I know what's happening here.  This is testing `NOT(case_property_missing(...))`, and `case_property_missing` is itself sort of a negative check, so we end up with a double negative.  If we cared to, we could probably instead `case_property_exists` and define `case_property_missing` as the reverse.  That'd likely give smaller queries, if that even matters.

However, this reminded me of this code that came up in the migration to ES2, where we automatically simplify logic because ES appeared incapable of handling certain constructs:
https://github.com/dimagi/commcare-hq/blob/104765a7e1be3be5e45f7f157f197e115eb57a66/corehq/apps/es/filters.py#L49-L65
It looks like we're switching from `or` to `should`, which doesn't trigger that simplification.

Reply:
Apparently ES7 does not have the same shortcoming as ES2 since it works even though that "simplification" is not applied to the ES7 query. I would vote to eliminate at least the first two branches (and possibly the third) of the `NOT` "simplification" in ES7. I think less automatic logic transformations will make it easier to troubleshoot queries.

However, this does make it a bit trickier write a function that automatically converts an ES7 query to an ES2 query (something I'm currently working on).

Reply:
Yes, this is happening due what Ethan mentioned above. There is also a [test-case](https://github.com/dimagi/commcare-hq/pull/28270/files/c5d65a8ff06e0b84819ea8378efa1cc5fb330500#diff-2cdb698195be0262401dd2863151ca53R513) that I added where the expression is tested on the data that makes the expected behaviour explicit..